### DISCUSSION TITLE
Finding the time step in the surface tracer flux

### DISCUSSION THREAD

**Discussion by francispoulin** [CliMA/Oceananigans.jl#2921] [General]
Building on the Ocean Wind Mixing and Convection example, we are trying to specify a surface temperature flux that depends on the time step.  If we have a fixed time step, this is easy and below is an example of that.  But if we use the time step wizard, which we want to, it becomes a bit harder.

Is there a way to find out the current time step in a function that we use to specify the tracer flux boundary conditions?

```
parameters = (cᴰ = 2.5e-3,
              ρₐ = 1.225,
              ρ₀ = 1030.0,        
              f  = coriolis.f,
              C  = 1.0,
              W  = 100,
              Cw = 3930,
              Δt = Δt)

@inline function T_flux(i, j, grid, clock, model_fields, p) 
        x   = xnode(Center(), i, grid)    
        t   = clock.time
        Lx  = grid.Lx
        dTH = p.W * p.Δt/(p.ρ₀ * p.Cw * grid.Δzᵃᵃᶜ)
        axx = exp.(-(8*x/Lx .- 8/2).^2)
        att = (p.C - cos(2π*t/1days)) 
        return dTH * axx * att    
    end
```

**Comment by glwagner** [comment]
One solution is to create a reference to the time-step:

```julia
Δt = Ref(Δt)
```

To access the value you write `p.Δt[]`.

Then the value must be updated in a callback function that captures the reference in a global variable:

```julia
update_Δt(sim) = Δt[] = sim.Δt
```

The callbacks are executed in the order they are added, so you need to update the time step _after_ the time step wizard is called:

```julia
simulation.callbacks[:wizard] = Callback(time_step_wizard, IterationInterval(10))
simulation.callbacks[:update_dt] = Callback(update_Δt, IterationInterval(10))
```

Make sure the callback that updates the time step and the callback for the wizard are called on the same schedule.

**Reply by francispoulin** [reply]
Thank you @glwagner !  That seems very clever and I will give it a try and let you know how it works.

**Reply by glwagner** [reply]
If you'd like a system that doesn't require capturing the reference as a global variable, you can use callback parameters (I think):

```julia
Δt_reference = Ref(Δt)
update_Δt(sim, Δt) = Δt[] = sim.Δt
simulation.callbacks[:update_dt] = Callback(update_Δt, IterationInterval(10), parameters=Δt_reference)
```
